File: storage/db-core/src/main/java/com/coffee/gu/cancel/CancelRepositoryImpl.java
package com.coffee.gu.cancel;

import com.coffee.gu.payment.Payment;
import org.springframework.stereotype.Repository;

import java.time.OffsetDateTime;

@Repository
public class CancelRepositoryImpl implements CancelRepository {

    private final CancelJpaRepository cancelJpaRepository;

    public CancelRepositoryImpl(CancelJpaRepository cancelJpaRepository) {
        this.cancelJpaRepository = cancelJpaRepository;
    }

    @Override
    public Cancel cancel(Payment payment, String externalCancelKey) {
        return cancelJpaRepository.save(
                new CancelEntity(
                        payment.getPrincipal().getKey(),
                        payment.getPrincipal().getType(),
                        payment.getOrderKey(),
                        payment.getId(),
                        payment.getOriginalAmount(),
                        payment.getIssuedCouponId(),
                        payment.getCouponDiscount(),
                        payment.getAmount(),
                        payment.getAmount(),
                        externalCancelKey,
                        OffsetDateTime.now()
                )
        ).toModel();
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/cancel/CancelRepositoryImpl.kt
package com.coffee.gu.cancel

import com.coffee.gu.payment.Payment
import org.springframework.stereotype.Repository
import java.time.OffsetDateTime

@Repository
class CancelRepositoryImpl(
    private val cancelJpaRepository: CancelJpaRepository,
) : CancelRepository {

    override fun cancel(payment: Payment, externalCancelKey: String): Cancel {
        return cancelJpaRepository.save(
            CancelEntity(
                principalKey = payment.principal.key,
                principalType = payment.principal.type,
                orderKey = payment.orderKey,
                paymentId = payment.id ?: 0L,
                originalAmount = payment.originalAmount,
                issuedCouponId = payment.issuedCouponId,
                couponDiscount = payment.couponDiscount,
                paidAmount = payment.amount,
                canceledAmount = payment.amount,
                externalCancelKey = externalCancelKey,
                canceledAt = OffsetDateTime.now(),
            ),
        ).toModel()
    }
}
